### PR TITLE
Feature:  Add Custom HTTP Headers and Method (via CLI and Programmatically)

### PR DESCRIPTION
**Pull request by ericzorn93**
This feature allows the user to send custom HTTP headers to the JS API in order to introspect and parse the remote URL Swagger/OpenAPI schema. This has been updated for the CLI format as well. These changes will not cause any breaking changes to the API and/or user.

All tests are passing.

### CODE REVIEW

**Comment by ericzorn93** [comment]
Many of these tests seem to be failing from a combination of failed network calls to introspect the schema and some indentation issues with type output in the format of strings.

**Comment by ericzorn93** [comment]
Hey @drwpow, it won't let me add reviewers to this PR for some reason.

**Comment by drwpow** [comment]
> Hey @drwpow, it won't let me add reviewers to this PR for some reason.

I believe you have to have repo access to do that. But feel free to just @- mention anyone, or I can add them myself.

File: tests/empty-definitions.test.ts
Comment (drwpow) [inline tests/empty-definitions.test.ts:34]: 👍🏻 I’m fine with snapshots here. As you may have guessed I’m partial to having the full-file “snapshots” just so they could be more-easily readable. Also in a project like this where nearly everything is a snapshot, I didn’t want a careless `-u` flag causing a huge regression and liked the added friction of having to edit a file.

But that’s for the full-schema tests; I agree that the inline snapshots here were noisy, and not as useful. I’m OK with this change.

**Comment by drwpow** [comment]
@ericzorn93 this is great! Very well-thought out, great execution, and thanks for improving the tests, too.

I’m happy to approve & merge this but let me know if you wanted additional reviewers as well.

**Comment by gr2m** [comment]
I'm sorry I don't have the bandwidth to review. But we gotta trust the tests. If something breaks, that's the fault of the test setup, not your PR :) Looks like great work, thank you!

**Comment by ericzorn93** [comment]
@gr2m Thank you. If we are good, do you or another maintainer mind merging?

**Comment by ericzorn93** [comment]
@drwpow Sorry to bother you again, but if this PR is safe to merge, do you mind merging to master? I think this would be a great help for my personal use case and others as well 👍

File: bin/cli.js
Comment (mbelsky) [inline bin/cli.js:37]: What do you think about passing header values as an array like curl does that? I guess it will be easier to write headers one by one in cli instead of making one json.

How curl does that: `curl -header "X-First-Name: John" --header "X-Last-Name: Doe" http://example.com/`

Also probably with this way we won't need to parse headers because `meow` could handle multiple values for a flag: https://github.com/sindresorhus/meow#flags

File: src/utils.ts
Comment (mbelsky) [inline src/utils.ts:152]: I found this function returns true for values like `123POST, POST123` etc. What about using [http.METHODS](https://nodejs.org/api/http.html#http_http_methods) to validate a value?

```suggestion
import {METHODS} from 'http'

export function isValidHTTPMethod(httpMethod: string | HTTPVerb): boolean {
  METHODS.includes(httpMethod.toUpperCase())
}
}
```

File: src/load.ts
Comment (mbelsky) [inline src/load.ts:161]: `typeof undefined !== "string "` so let's simplify this condition a bit:
```suggestion
      if (typeof options.httpMethod === "string" && isValidHTTPMethod(httpMethod)) {
```

File: src/load.ts
Comment (mbelsky) [inline src/load.ts:147]: What about using `Object.entries` here instead of `Object.keys`?

**Review by mbelsky — CHANGES_REQUESTED** [review]
This is a great feature that I'd really would like to have in this tool.

File: src/load.ts
Comment (ericzorn93) [reply src/load.ts:147]: I know that we can access the value using the entries method, but is there a preference of why we would do this?￼

File: src/load.ts
Comment (ericzorn93) [reply src/load.ts:161]: Fair enough, this was an oversight on my end, thank you for pointing this out.￼

File: bin/cli.js
Comment (ericzorn93) [reply bin/cli.js:37]: Is this a block or to getting merged? Or can we add this at a later point?￼￼

File: src/utils.ts
Comment (ericzorn93) [reply src/utils.ts:152]: Great point! I will fix this as well￼

File: bin/cli.js
Comment (ericzorn93) [reply bin/cli.js:37]: I definitely feel that this is a very useful feature and it has a way cleaner API, but can this be an additional option? I think both ways would be valid ways of inputting HTTP headers￼￼

Comment (ericzorn93) [reply bin/cli.js:37]: @mbelsky I just realized a unique edge case on why this might not be the best API for the user. Currently, the values are the only thing stored in the array. Therefore, we would have to enforce a strict string schema of ‘-x "headerKey:headerValue‘ and then split on the string. However, compared to a JSON object (yes, definitely more verbose), they will potentially run into issues with spacing and casing. We would need much more strict validation

File: src/load.ts
Comment (ericzorn93) [reply src/load.ts:160]: Complete

File: src/load.ts
Comment (ericzorn93) [reply src/load.ts]: Completed

File: bin/cli.js
Comment (ericzorn93) [reply bin/cli.js:37]: @drwpow What do you think about this implementation? I personally think that we should be okay to merge and then think about this additional feature and implementation next.

Comment (mbelsky) [reply bin/cli.js:37]: We have [discussed](https://github.com/drwpow/openapi-typescript/issues/645#issuecomment-868938899) with @drwpow about this feature. Just wanna to link the issue to this pr.

File: src/load.ts
Comment (mbelsky) [reply src/load.ts]: There is only one preference why I've asked about this: we rarely use the classic `for (let i=0; i<…)` statement to iterate a list without making any index magic because we have better version of that: for in, for of, forEach. These modern versions are less verbose and more semantically clear. So same point for `Object.entries`.

File: bin/cli.js
Comment (ericzorn93) [reply bin/cli.js:37]: I can add this as an additional alternative￼

File: .nvmrc
Comment (gr2m) [inline .nvmrc:1]: This change seems out of scope of this pull request

**Comment by gr2m** [comment]
Once you two are in agreement and confirmed that it's working, can you please ping Drew? It's up to him whether the added complexity is something he wants to maintain in the long term. You have my blessing, but I'm just helping out :)

**Comment by gr2m** [comment]
sounds like you'll have to bump the coverage ☝🏼 it shouldn't drop

File: .nvmrc
Comment (ericzorn93) [reply .nvmrc:1]: Np, this can be removed. I was trying to enforce consistency but this should not be an issue for most people using Node 14>=

File: bin/cli.js
Comment (ericzorn93) [reply bin/cli.js:44]: @mbelsky and @drwpow We now support both methods. Please check the updated PR.

File: .nvmrc
Comment (ericzorn93) [reply .nvmrc]: Removed :)

**Comment by ericzorn93** [comment]
@gr2m I have added a number of small tests for utility functions to attempt to bump coverage. I don't know how much more I can really increase unit test coverage of the load function without adding insignificant granular unit tests

**Comment by ericzorn93** [comment]
@gr2m all test coverage has been updated

**Comment by ericzorn93** [comment]
- [x] Hey @gr2m, @mbelsky, and @drwpow. Sorry to be pushy, but my PR should be good to go with all requested changes and increased test coverage. Now, you can add headers individually like the CURL CLI, or via a JSON string. 

Do you mind merging?

**Comment by gr2m** [comment]
> This functionality would be great in my current project 🎉

To take the pressure of, maybe publish your own fork for the time being? I don't know when @drwpow will have the time to review, and it's bad etiquette in independent Open Source projects to pressure maintainers like that, please don't do that.

**Comment by ericzorn93** [comment]
@gr2m I completely understand and did not think of that tbh. I will do this in the meantime. I apologize for any pressure I have put on the maintainers

**Comment by drwpow** [comment]
@ericzorn93 sorry for the delay on this! This library has been difficult for me to maintain the past few months. I went ahead and approved and merged this in #764, with a few changes as-noted. See that PR for more details.

This approach of merging your work in another PR is a bit unorthodox, and I apologize. Please check that off as a failure on my part to give you a timely review! If you’d like to adjust anything from my revisions, please open up another PR.

Overall I wanted to say again **what a quality PR this was!** Your JS Docs are excellent, your comments are excellent, and I am amazed at the amount of thought & care you put into this PR. And I apologize again for not merging this in sooner because it is quality work.

I’ll be shipping a new minor version with this feature soon!

**Comment by drwpow** [comment]
@all-contributors please add @ericzorn93 for code, test, docs

**Comment by ericzorn93** [comment]
@drwpow No worries at all. I completely understand. I apologize for coming across pushy, I was just trying to incorporate this on a script that I have been working on. 

I completely understand that maintaining widely used open source software can be a difficult task and very time consuming. This is a great repo!

Thanks for the kind words and I look forward to contributing more in the future 😀.

Thanks,
Eric

**Pull request closed without merging**
